Check that actual files are modified by pipelines

nwiltsie opened the pull request:
# Description
This adds a check that the `actual` files are have a modification time after the creation of their associated `NFTestAssert`. Here's an example from https://github.com/uclahs-cds/pipeline-filter-RNAEditingSite/pull/13 showing how I was comparing to an existing-but-incorrect file:

```
2023-09-26 22:07:30,262 - NextFlow - INFO - All positions: 74
2023-09-26 22:07:30,262 - NextFlow - INFO - Positions filtered in: 0
2023-09-26 22:07:30,396 - NextFlow - INFO - Positions filtered out: 74
2023-09-26 22:07:30,396 - NextFlow - INFO -
2023-09-26 22:07:30,397 - NextFlow - INFO -
2023-09-26 22:07:30,476 - NFTest - DEBUG - Test creation time: 2023-09-26 22:07:00.532961+00:00
2023-09-26 22:07:30,476 - NFTest - DEBUG - Actual mod time:    2023-09-01 21:21:11.930667+00:00
2023-09-26 22:07:30,476 - NFTest - ERROR - ('/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie_standardize/A-mini-n2/filter-RNAEditingSite-1.0.0/CPCG0196-F1-A-mini-n2/REDItools-1.3/output/CPCG0196-F1-A-mini-n2_annotated.txt was not modified by this pipeline',)
2023-09-26 22:07:30,476 - NFTest - ERROR -  [ failed ]
```

Here's an example where I _did_ have that `-rc` appended to the pipeline version:
```
2023-09-26 22:04:09,510 - NextFlow - INFO - All positions: 74
2023-09-26 22:04:09,510 - NextFlow - INFO - Positions filtered in: 0
2023-09-26 22:04:09,510 - NextFlow - INFO - Positions filtered out: 74
2023-09-26 22:04:09,510 - NextFlow - INFO -
2023-09-26 22:04:09,510 - NextFlow - INFO -
2023-09-26 22:04:09,559 - NFTest - DEBUG - Test creation time: 2023-09-26 22:03:41.858707+00:00
2023-09-26 22:04:09,560 - NFTest - DEBUG - Actual mod time:    2023-09-26 22:04:00.621141+00:00
2023-09-26 22:04:09,560 - NFTest - DEBUG - md5 /hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie_standardize/A-mini-n2/filter-RNAEditingSite-1.0.0-rc/CPCG0196-F1-A-mini-n2/REDItools-1.3/output/CPCG0196-F1-A-mini-n2_annotated.txt /hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/output/REDItools-1.3/output/CPCG0196-F1-A-mini-n2_annotated.txt
2023-09-26 22:04:09,561 - NFTest - DEBUG - Assertion passed
```

# Checklist
<!--- Please read each of the following items and confirm by replacing the [ ] with a [X] --->

- [x] This PR **does *NOT* contain** Protected Health Information [(PHI)](https://ohrpp.research.ucla.edu/hipaa/). A repo may ***need to be deleted*** if such data is uploaded. <br> Disclosing PHI is a ***major problem***[^1] - Even ***a small leak can be costly***[^2].
  
- [x] This PR **does *NOT* contain** germline genetic data[^3], RNA-Seq, DNA methylation, microbiome or other molecular data[^4].

[^1]: [UCLA Health reaches $7.5m settlement over 2015 breach of 4.5m patient records](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m)
[^2]: [The average healthcare data breach costs $2.2 million, despite the majority of breaches releasing fewer than 500 records.](https://www.ponemon.org/local/upload/file/Sixth%20Annual%20Patient%20Privacy%20%26%20Data%20Security%20Report%20FINAL%206.pdf)
[^3]: [Genetic information is considered PHI.](https://www.genome.gov/about-genomics/policy-issues/Privacy#:~:text=In%202013%2C%20as%20required%20by,genetic%20information%20for%20underwriting%20purposes.)
  [Forensic assays can identify patients with as few as 21 SNPs](https://www.sciencedirect.com/science/article/pii/S1525157817305962)
[^4]: [RNA-Seq](https://www.nature.com/articles/ng.2248), [DNA methylation](https://ieeexplore.ieee.org/document/7958619), [microbiome](https://www.pnas.org/doi/pdf/10.1073/pnas.1423854112), or other molecular data can be used to predict genotypes (PHI) and reveal a patient's identity.


- [x] This PR **does *NOT* contain** other non-plain text files, such as: compressed files, images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other output files.

_&emsp; To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example._

- [x] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [x] I have set up or verified the `main` branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [x] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
  
- [x] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

yashpatel6 reviewed (commented):
This is great! A couple of things: when a timestamp issue is found, does NFTest stop running all tests? The runner I believe is only specifically catching AssertionError: https://github.com/uclahs-cds/tool-NFTest/blob/ef3b0eb4609e99c9d0462e346c8c2a944b57e911/nftest/NFTestRunner.py#L45-L53

It'd be worth updating the tests to include case(s) for the additions here

nwiltsie commented:
That's a great catch - you can see in my first log snippet that it does halt immediately, which is a change from the current behavior. I'll just switch to using a standard `assert` rather than the unnecessary new exception type.

yashpatel6 commented on nftest/NFTestAssert.py:
```diff
@@ -34,6 +38,16 @@ def assert_expected(self):
+        file_mod_time = datetime.datetime.fromtimestamp(
+            Path(self.actual).stat().st_mtime,
+            tz=datetime.timezone.utc
+
```
Extra line here?

nwiltsie replied:
Oops! Fixed.

yashpatel6 reviewed (commented):
Looks good! It'd be useful to add a test case for the assertion here being properly thrown/caught

nwiltsie commented:
I've fixed the tests (again - I keep forgetting to check them) and added suitable mocks to demonstrate that the file modification times can cause the expected asserts.

yashpatel6 commented on test/unit/test_NFTestAssert.py:
```diff
@@ -35,32 +38,56 @@ def test_get_assert_method_method(mock_assert):
     mock_assert.return_value.mock_assert_expected = NFTestAssert.assert_expected
 
     assert mock_assert().mock_assert_expected(mock_assert()) is None
+
+    # Set the test start time to be after the timestamp
+    mock_assert.return_value.startup_time = datetime.datetime.fromtimestamp(
+        timestamp+1,
+        tz=datetime.timezone.utc
+    )
+
+    with pytest.raises(AssertionError):
+        mock_assert().mock_assert_expected(mock_assert())
```
It may be worth putting this into a separate test case since the test here intended to pass the assertions, whereas the addition is testing for proper failed assertion based on timestamp difference. It can probably be done through parametrize-ing the different settings in `test_assert_expected_fail` to avoid duplicating the entire test, like here:
https://github.com/uclahs-cds/package-PipeVal/blob/997f7d2bdb5e3e092a1b8369dedb5439387a97c4/test/unit/test_generate_checksum.py#L49-L59

nwiltsie replied:
Yeah really it should be a slightly larger refactor so that we can test the equality independently of the timestamps... I'll rework it.

nwiltsie commented:
@yashpatel6 I think I have the tests in a better place, but this is getting a little sprawling. Let me know if you'd prefer that I open a clean PR with the refactoring and testing changes before dealing with the new feature.

yashpatel6 commented:
> @yashpatel6 I think I have the tests in a better place, but this is getting a little sprawling. Let me know if you'd prefer that I open a clean PR with the refactoring and testing changes before dealing with the new feature.

Are there any more changes you'd want to add to this PR? It should be in a good place at the moment in terms of having the tests in place to proceed and we can have a separate PR to clean up tests if necessary

nwiltsie commented:
No, this is everything I'm intending to touch! I'm happy to merge if you're good with these changes.

yashpatel6 reviewed (approved):
Looks good!
